File: 0221-Maximal-Square/py-0221/Solution2.py
# Source : https://leetcode.com/problems/maximal-square/
# Author : penpenps
# Time   : 2019-07-24

from typing import List

# modify matrix in-place
# Time Complexity: O(n^2)
# Space Complexity: O(1)

class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        if not matrix or not matrix[0]: return 0
        n = len(matrix)
        m = len(matrix[0])
        
        for i in range(n):
            matrix[i][0] = 1 if matrix[i][0] == '1' else 0 
        for j in range(1, m):
            matrix[0][j] = 1 if matrix[0][j] == '1' else 0 
        
        for i in range(1, n):
            for j in range(1, m):
                if matrix[i][j] == '1':
                    matrix[i][j] = min(matrix[i-1][j-1], matrix[i][j-1], matrix[i-1][j]) + 1
                else:
                    matrix[i][j] = 0
        return max(max(row) for row in matrix)**2

File: 0001-0500/0219-Contains-Duplicate-II/java-0219/src/Solution2.java
/// Source : https://leetcode.com/problems/contains-duplicate-ii/description/
/// Author : liuyubobobo
/// Time   : 2019-04-10

import java.util.HashMap;

/// Using Hash Map
/// Time Complexity: O(n)
/// Space Complexity: O(n)
public class Solution2 {

    public boolean containsNearbyDuplicate(int[] nums, int k) {

        if(nums == null || nums.length <= 1)
            return false;

        if(k <= 0)
            return false;

        HashMap<Integer, Integer> record = new HashMap<>();
        record.put(nums[0], 0);
        for(int i = 1; i < nums.length; i ++){
            if(record.containsKey(nums[i]) && i - record.get(nums[i]) <= k)
                return true;

            record.put(nums[i], i);
        }

        return false;
    }

    private static void printBool(boolean b){
        System.out.println(b ? "True" : "False");
    }

    public static void main(String[] args) {

        int[] nums = {1, 2, 1};
        int k = 1;
        printBool((new Solution()).containsNearbyDuplicate(nums, k));
    }
}
